filters/openpolicyagent/openpolicyagent.go

```diff
@@ -0,0 +1,524 @@
+
+	for key, inst := range registry.instances {
+		lastused, ok := registry.lastused[inst]
+
+		if ok && t.Sub(lastused) > registry.reuseDuration {
+			inst.Close(ctx)
+
+			delete(registry.instances, key)
+			delete(registry.lastused, inst)
+		}
+	}
+}
+
+func (registry *OpenPolicyAgentRegistry) startCleanerDaemon() {
+	ticker := time.NewTicker(10 * time.Second)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-registry.quit:
+			return
+		case t := <-ticker.C:
+			registry.cleanUnusedInstances(t)
+		}
+	}
+}
+
+// Do implements routing.PostProcessor and cleans unused OPA instances
+func (registry *OpenPolicyAgentRegistry) Do(routes []*routing.Route) []*routing.Route {
+	inUse := make(map[*OpenPolicyAgentInstance]struct{})
+
+	for _, ri := range routes {
+		for _, fi := range ri.Filters {
+			if ff, ok := fi.Filter.(OpenPolicyAgentFilter); ok {
+				inUse[ff.OpenPolicyAgent()] = struct{}{}
+			}
+		}
+	}
+
+	registry.markUnused(inUse)
+
+	return routes
+}
+
+func (registry *OpenPolicyAgentRegistry) NewOpenPolicyAgentInstance(bundleName string, config OpenPolicyAgentInstanceConfig, filterName string) (*OpenPolicyAgentInstance, error) {
+	registry.mu.Lock()
+	defer registry.mu.Unlock()
+
+	if registry.closed {
+		return nil, fmt.Errorf("open policy agent registry is already closed")
+	}
+
+	if instance, ok := registry.instances[bundleName]; ok {
+		delete(registry.lastused, instance)
+
+		return instance, nil
+	}
+
+	instance, err := registry.newOpenPolicyAgentInstance(bundleName, config, filterName)
+
+	if err != nil {
+		return nil, err
+	}
+
+	registry.instances[bundleName] = instance
+
+	return instance, nil
+}
+
+func (registry *OpenPolicyAgentRegistry) markUnused(inUse map[*OpenPolicyAgentInstance]struct{}) error {
+	registry.mu.Lock()
+	defer registry.mu.Unlock()
+
+	for _, instance := range registry.instances {
+		if _, ok := inUse[instance]; !ok {
+			registry.lastused[instance] = time.Now()
+		}
+	}
+
+	return nil
+}
+
+func (registry *OpenPolicyAgentRegistry) newOpenPolicyAgentInstance(bundleName string, config OpenPolicyAgentInstanceConfig, filterName string) (*OpenPolicyAgentInstance, error) {
+	runtime.RegisterPlugin(envoy.PluginName, envoy.Factory{})
+
+	configBytes, err := interpolateConfigTemplate(config.configTemplate, bundleName)
+	if err != nil {
+		return nil, err
+	}
+
+	engine, err := New(inmem.New(), configBytes, config, filterName, bundleName)
+
+	if err != nil {
+		return nil, err
+	}
+
+	ctx := context.Background()
+	if err = engine.Start(ctx); err != nil {
+		return nil, err
+	}
+
+	err = engine.waitPluginsReady(100*time.Millisecond, 30*time.Second)
+	if err != nil {
+		engine.Logger().WithFields(map[string]interface{}{"err": err}).Error("Failed to wait for plugins activation.")
+		return nil, err
+	}
+
+	return engine, nil
+}
+
+type OpenPolicyAgentInstance struct {
+	manager        *plugins.Manager
+	instanceConfig OpenPolicyAgentInstanceConfig
+	opaConfig      *config.Config
+	bundleName     string
+
```

Comment:
is this whitespace forming a group? If so, what kind of group (maybe document it with a comment)?

Reply:
It is actually not any longer, will remove the whitespace